tools/twitter-api/twitter_api_handler.py

```diff
@@ -0,0 +1,86 @@
+from scrolling import es, es_index, search
+from elasticsearch import Elasticsearch
+from error_handling import *
+from elasticsearch.helpers import bulk
+from random import randint
+import json
+import tweepy
+
+class TwitterAPIHandler:
+  def __init__(self):
+    creds = json.load(open("creds.json"))
+    self.creds = creds["twitter-creds"]
+    
+    self.consumer_key        = self.creds["consumer_key"]
+    self.consumer_key_secret = self.creds["consumer_key_secret"]
+    self.access_token        = self.creds["access_token"]
+    self.access_token_secret = self.creds["access_token_secret"]
+    self.auth = tweepy.OAuthHandler(self.consumer_key, self.consumer_key_secret)
+    self.auth.set_access_token(self.access_token, self.access_token_secret)
+    self.api = tweepy.API(self.auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)
+
+  def GetElasticSearchHitsWithScrolling(self, max_hits):
+    """
+    Uses Thomas Shweh's ElasticSearch scrolling algorithm to get max_hits number of Tweets from
+    ElasticSearch
+    """
+    return search(max_hits=max_hits-1000)
+      
+  def GetFullText(self, apiResponse, index):
+    """
+    For testing to see if the tweet has an extended text
+    """
+    try:
+      return apiResponse[index].full_text
+    except:
+      return -1
+    
+
+  def GetTweetsFromAPI(self, hits):
+    """
+    Gets the original tweet for each response-type hit from Twitter 
+    and writes that original tweet to the exisiting data
+    """
+    output = list()
+    responseTweetIds = [hit['_source']['in_reply_to_status_id'] for hit in hits]
+    numUnavailable = 0
+    for i in range(0, len(responseTweetIds), 100):
+      apiResponse = self.api.statuses_lookup(responseTweetIds[i:i+100], tweet_mode="extended", map=True)
+      responseIndex = 0
+      for j in range(i, i+100):
+        currHit = hits[j]
+        currHit['_source']['in_reply_to_status'] = apiResponse[responseIndex]
+          
+        responseIndex += 1
+        
+        output.append(currHit)
+        
+    return output, numUnavailable
+  
+  def WriteDataToElasticSearch(self, hits):
+    """
+    Write the modified data back to the elasticsearch index
+    """
+    for i in range(0, len(hits), 100):
+      process_batch = hits[i:i+100]
```

Comment:
The Elasticsearch bulk API requires a wrapper around each document (tweet) json to indicate what Elasticsearch should do with it. 

See https://github.com/TheRensselaerIDEA/COVID-Twitter/blob/0245a22ad7fd187db028c93d5dbee2ac53feb9d2/twitter_monitor/jsonl_dataset_loader.py#L82

In this case since we are writing back to an existing document, we would want to set _op_type to "update" (reference: https://www.elastic.co/guide/en/elasticsearch/reference/7.9/docs-bulk.html#bulk-api-request-body)